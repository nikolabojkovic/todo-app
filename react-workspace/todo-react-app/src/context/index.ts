export * from './IState';
export * from './TodoListContext';
export * from './testData';