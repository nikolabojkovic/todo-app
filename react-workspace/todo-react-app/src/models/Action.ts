export interface IAction {
  type: string,
  payload: any
}

export enum TodoActions {
  fetch = 'fetch',
  filter = 'filter',
  search = 'search',
  sort = 'sort',

  fetched = 'fetched',
  filtered = 'filtered',
  sorted = 'sorted',
  searched = 'searched',
  imported = 'imported',
  loadingStarted = 'loading-started',

  searchTermUpdated = 'searchTerm-updated',
  pagingUpdated = 'paging-updated',

  added = 'added',
  changed = 'changed',
  deleted = 'deleted',
}