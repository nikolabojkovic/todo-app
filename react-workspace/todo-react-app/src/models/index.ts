export * from './Action';
export * from './Event';
export * from './IFilter';
export * from './IPaging';
export * from './ISearch';
export * from './ISettings';
export * from './ISort';
export * from './Todo';