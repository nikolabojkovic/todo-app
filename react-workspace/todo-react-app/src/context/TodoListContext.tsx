import { createContext, useContext, useReducer } from "react";

import { IAction, TodoActions, IPaging, ITodo, StateFilter, ISettings } from "../models";
import { IState } from "./";
 
export const TodosContext = createContext({} as IState);
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const TodosDispatchContext = createContext(null as any);

type Props = {
  children: JSX.Element | JSX.Element[],
  initialState: IState
}

export function TodoStateProvider({ initialState, children }: Props) {
  const [state, dispatch] = useReducer(
    todoStateReducer,
    initialState
  );

  return (
    <TodosContext.Provider value={state}>
      <TodosDispatchContext.Provider value={dispatch}>
        {children}
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>
  );
}

export function useTodoList() {
  return useContext(TodosContext);
}

export function useTodoListDispatch() {
  return useContext(TodosDispatchContext);
}

export default TodosDispatchContext.Provider;

export function todoStateReducer(state: IState, action: IAction) {
  switch (action.type) { 
    case TodoActions.loadingStarted: {
      return { 
        ...state,
        effectTrigger: null,
        isLoading: true
      } as IState;
    }
    case TodoActions.activeTabChanged: {
      return { 
        ...state,
        activeTab: action.payload.activeTab
      } as IState;
    }
    case TodoActions.fetch: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.fetch, payload: action.payload },
      } as IState;
    }
    case TodoActions.filter: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.filter, payload: action.payload },
      } as IState;
    }
    case TodoActions.search: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.search, payload: action.payload },
      } as IState;
    }
    case TodoActions.sort: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.sort, payload: action.payload },
      } as IState;
    }

    case TodoActions.fetched: {
      return {
        ...state,
        isLoading: false,
        originalList: action.payload.list,
        displayList: action.payload.list,
        effectTrigger: null,
        paging: {
          ...state.paging,
          totalCount: action.payload.list.length,
          activePage: state.paging.activePage === 0 ? 1 : state.paging.activePage,
          itemsPerPage: state.paging.itemsPerPage,
          startIndex: ((state.paging.activePage === 0 ? 1 : state.paging.activePage) - 1) * state.paging.itemsPerPage,
          endIndex: (state.paging.activePage === 0 ? 1 : state.paging.activePage) * state.paging.itemsPerPage
        } as IPaging,
      };
    }
    case TodoActions.searched: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.filtered: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        filter: {...action.payload.filter},
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.sorted: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        sort: {...action.payload.sort},
        paging: {...state.paging} as IPaging
      };
    }
    case TodoActions.imported: {
      return {
        ...state,
        originalList: [...action.payload.list],
        displayList: [...action.payload.list],
        effectTrigger: { type: TodoActions.imported },
        search: { searchTerm: '' },
        filter: {
          state: StateFilter.all
        },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }

    case TodoActions.searchTermUpdated: {
      return {
        ...state,
        effectTrigger: null,
        search: { searchTerm: action.payload.searchTerm },
      };
    }    
    case TodoActions.pagingFatched: {
      return {
        ...state,
        effectTrigger: null,
        paging: { 
          ...state.paging,
          totalCount: action.payload.totalCount,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.pagingUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.pagingUpdated },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.settingsFetched: {
      return {
        ...state,
        effectTrigger: null,
        settings: {
          ...action.payload
        } as ISettings
      };
    }
    case TodoActions.settingsUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.settingsUpdated },
        settings: {
          ...action.payload
        } as ISettings
      };
    }

    case TodoActions.added: {
      const id = state.originalList.length >= 1 
      ? state.originalList
          .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[state.originalList.length - 1].id 
      : 0;
      const newTodo = {
        id: id + 1,
        title: action.payload.title,
        description: action.payload.description,
        completed: false, 
        createdAt: action.payload.createdAt
      } as ITodo;

      return {
        ...state,
        originalList: [...state.originalList, newTodo],
        displayList: [...state.displayList, newTodo],
        effectTrigger: { type: TodoActions.added },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount + 1,
          activePage: calculateActivePageOnAdd(state.paging),
          startIndex: (calculateActivePageOnAdd(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnAdd(state.paging) * state.paging.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.changed: {
      return {
        ...state,
        originalList: state.originalList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        displayList: state.displayList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        effectTrigger: { type: TodoActions.changed },
        paging: {...state.paging}
      } as IState;
    }
    case TodoActions.deleted: {
      return {
        ...state,
        originalList: state.originalList.filter(t => t.id !== action.payload.id),
        displayList: state.displayList.filter(t => t.id !== action.payload.id),
        effectTrigger: { type: TodoActions.deleted },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount - 1,
          activePage: calculateActivePageOnDelete(state.paging),
          startIndex: (calculateActivePageOnDelete(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnDelete(state.paging) * state.paging.itemsPerPage,  
        } as IPaging
      } as IState;
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

function calculateActivePageOnDelete(paging: IPaging) {
  return Math.ceil((paging.totalCount - 1) / paging.itemsPerPage) < paging.activePage
    ? paging.activePage - 1 : paging.activePage;
}

function calculateActivePageOnAdd(paging: IPaging) {
  return Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) > paging.activePage
    ? Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) : paging.activePage;
}