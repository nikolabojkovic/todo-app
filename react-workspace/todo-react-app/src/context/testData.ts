import { ITodo } from "../models/Todo";
import { IPaging } from "../models/IPaging";
import { IState } from "./IState";
import { ISort, SortDirection } from "../models/ISort";
import { IFilter, StateFilter } from "../models/IFilter";
import { IGeneralSettings, IPaginationSettings, ISearchSettings, ISettings, ListContainerType, PaginationType } from "../models/ISettings";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings
  } as ISettings
} as IState;