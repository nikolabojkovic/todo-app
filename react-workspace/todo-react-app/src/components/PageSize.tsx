import { Stack, Col } from 'react-bootstrap';
import Dropdown from 'react-bootstrap/Dropdown';
import { useTodoList, useTodoListDispatch } from '../context/TodoListContext';
import { IAction } from '../models/Action';
import { useState } from 'react';

type Props = {
  inputSelectRef: any,
  pageCount: number
};

export function PageSize({ inputSelectRef, pageCount }: Props) {

  const [itemsPerPage, setItemsPerPage] = useState(5);

  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function handlePageSizeChange(pageSize: number) {
    dispatch({
      type: 'paging-updated',
      payload: {
        activePage: 1,
        itemsPerPage: pageSize
      }
    } as IAction);
    setItemsPerPage(pageSize);
  }
  
  return (
  <>
    <Col sm={3} className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center">
      Page {todoList.paging.activePage} of {pageCount}
    </Col>
    <Col sm={4} className="p-3 pt-2 pb-2">
      <Stack direction="horizontal" gap={2} className="justify-content-center">
        <div className='ms-sm-auto'>
          Page size:{' '}
        </div>
        <div className="">
          <Dropdown>
            <Dropdown.Toggle
              className="action-button" 
              ref={inputSelectRef} 
              id="dropdown-basic"
            >
              {' '}{itemsPerPage}{' '}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item href="#/action-1" onClick={() => handlePageSizeChange(5)}>5</Dropdown.Item>
              <Dropdown.Item href="#/action-2" onClick={() => handlePageSizeChange(10)}>10</Dropdown.Item>
              <Dropdown.Item href="#/action-3" onClick={() => handlePageSizeChange(50)}>50</Dropdown.Item>
              <Dropdown.Item href="#/action-3" onClick={() => handlePageSizeChange(100)}>100</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </Stack>
    </Col>
  </>)
}