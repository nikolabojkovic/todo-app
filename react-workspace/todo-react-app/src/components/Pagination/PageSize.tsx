import { Stack, Col } from 'react-bootstrap';
import Dropdown from 'react-bootstrap/Dropdown';
import { useTodoListDispatch } from '../../context/TodoListContext';
import { IAction, TodoActions } from '../../models/Action';

type Props = {
  inputSelectRef: ((instance: HTMLButtonElement | null) => void) | React.RefObject<HTMLButtonElement> | null | undefined,
  pageCount: number,  
  activePage: number,
  totalCount: number,
  pageSize?: number | null | undefined
};

export function PageSize({ inputSelectRef, pageCount, activePage, totalCount, pageSize = 5 }: Props) {
  const dispatch = useTodoListDispatch();

  function handlePageSizeChange(pageSize: number) {
    dispatch({
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: pageSize
      }
    } as IAction);
  }

  return (
  <>
    <Col sm={3} className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center">
      Page {activePage} of {pageCount}
    </Col>
    <Col sm={4} className="p-3 pt-2 pb-2">
      <Stack direction="horizontal" gap={2} className="justify-content-center">
        <div className='ms-sm-auto'>
          Page size:{' '}
        </div>
        <div className="">
          <Dropdown>
            <Dropdown.Toggle
              data-testid="selected-size-option"
              variant="outline-secondary"
              className="action-button" 
              ref={inputSelectRef} 
              id="dropdown-basic"
              disabled={totalCount === 0}
            >
              {' '}{pageSize}{' '}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item data-testid="size-option-5" href="#/action-1" onClick={() => handlePageSizeChange(5)}>5</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-10" href="#/action-2" onClick={() => handlePageSizeChange(10)}>10</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-50" href="#/action-3" onClick={() => handlePageSizeChange(50)}>50</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-100" href="#/action-3" onClick={() => handlePageSizeChange(100)}>100</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </Stack>
    </Col>
  </>);
}