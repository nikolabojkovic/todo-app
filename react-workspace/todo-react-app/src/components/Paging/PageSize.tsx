import { Stack, Col } from 'react-bootstrap';
import Dropdown from 'react-bootstrap/Dropdown';
import { useTodoList, useTodoListDispatch } from '../../context/TodoListContext';
import { IAction, TodoActions } from '../../models/Action';
import { useState } from 'react';

type Props = {
  inputSelectRef: any,
  pageCount: number,
  pageSize: number
};

export function PageSize({ inputSelectRef, pageCount, pageSize = 5 }: Props) {

  const [itemsPerPage, setItemsPerPage] = useState(pageSize);

  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function handlePageSizeChange(pageSize: number) {
    dispatch({
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: pageSize
      }
    } as IAction);
    setItemsPerPage(pageSize);
  }
  
  return (
  <>
    <Col sm={3} className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center">
      Page {todoList.paging.activePage} of {pageCount}
    </Col>
    <Col sm={4} className="p-3 pt-2 pb-2">
      <Stack direction="horizontal" gap={2} className="justify-content-center">
        <div className='ms-sm-auto'>
          Page size:{' '}
        </div>
        <div className="">
          <Dropdown>
            <Dropdown.Toggle
              data-testid="selected-size-option"
              variant="outline-secondary"
              className="action-button" 
              ref={inputSelectRef} 
              id="dropdown-basic"
              disabled={todoList.paging.totalCount === 0}
            >
              {' '}{itemsPerPage}{' '}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item data-testid="size-option-5" href="#/action-1" onClick={() => handlePageSizeChange(5)}>5</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-10" href="#/action-2" onClick={() => handlePageSizeChange(10)}>10</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-50" href="#/action-3" onClick={() => handlePageSizeChange(50)}>50</Dropdown.Item>
              <Dropdown.Item data-testid="size-option-100" href="#/action-3" onClick={() => handlePageSizeChange(100)}>100</Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </Stack>
    </Col>
  </>)
}