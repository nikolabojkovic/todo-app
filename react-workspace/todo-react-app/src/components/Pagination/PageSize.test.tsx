import renderer, { act } from 'react-test-renderer';
import { PageSize } from './PageSize';
import { TodoStateProvider, TodosContext, TodosDispatchContext } from '../../context/TodoListContext';
import { stateTestData } from '../../context/testData';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IAction, TodoActions } from '../../models/Action';

describe('PageSize', () => {
  const globalContext = {
    state: {
      ...stateTestData,
    },
    dispatch: jest.fn()
  };
  const inputSelectRef = {
    current: {
      focus: jest.fn()
    }
  } as unknown as React.RefObject<HTMLButtonElement> | ((instance: HTMLButtonElement | null) => void) | null | undefined;

  it('component should match snapshot page size 10', () => {
    const jsxElement =
      (<TodoStateProvider initialState={stateTestData}>
        <PageSize inputSelectRef={inputSelectRef} pageCount={1} pageSize={10} totalCount={6} activePage={1} />
      </TodoStateProvider>);
    const tree = renderer.create(jsxElement).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display default page size', () => {
    const jsxElement =
      (<TodoStateProvider initialState={stateTestData}>
        <PageSize inputSelectRef={inputSelectRef} pageCount={1} totalCount={6} activePage={1} />
      </TodoStateProvider>);
    const tree = renderer.create(jsxElement).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should select page size 5', async () => {
    render(
      <TodosContext.Provider value={globalContext.state}>
        <TodosDispatchContext.Provider value={globalContext.dispatch} >
          <PageSize inputSelectRef={inputSelectRef} pageCount={1} pageSize={10} totalCount={6} activePage={1} />
        </TodosDispatchContext.Provider>
      </TodosContext.Provider>);
    const actionPagingUpdated = {
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: 5
      }
    } as IAction;

    let selectedFilterOption = screen.getByTestId('selected-size-option');

    await act(async () => {      
      return await userEvent.click(selectedFilterOption);    
    });

    const filterDropdown = screen.getByTestId('size-option-5');
      
    fireEvent.click(filterDropdown);  
    selectedFilterOption = screen.getByTestId('selected-size-option');

    expect(globalContext.dispatch).toBeCalledWith(actionPagingUpdated);
  });

  it('should select page size 10', async () => {
    render(
      <TodosContext.Provider value={globalContext.state}>
        <TodosDispatchContext.Provider value={globalContext.dispatch} >
          <PageSize inputSelectRef={inputSelectRef} pageCount={2} pageSize={5} totalCount={6} activePage={1} />
        </TodosDispatchContext.Provider>
      </TodosContext.Provider>);
    const actionPagingUpdated = {
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: 10
      }
    } as IAction;

    let selectedFilterOption = screen.getByTestId('selected-size-option');

    await act(async () => {      
      return await userEvent.click(selectedFilterOption);    
    });

    const filterDropdown = screen.getByTestId('size-option-10');
      
    fireEvent.click(filterDropdown);  
    selectedFilterOption = screen.getByTestId('selected-size-option');

    expect(globalContext.dispatch).toBeCalledWith(actionPagingUpdated);
  });

  it('should select page size 50', async () => {
    render(
      <TodosContext.Provider value={globalContext.state}>
        <TodosDispatchContext.Provider value={globalContext.dispatch} >
          <PageSize inputSelectRef={inputSelectRef} pageCount={2} pageSize={5} totalCount={6} activePage={1} />
        </TodosDispatchContext.Provider>
      </TodosContext.Provider>);
    const actionPagingUpdated = {
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: 50
      }
    } as IAction;
    
    let selectedFilterOption = screen.getByTestId('selected-size-option');

    await act(async () => {      
      return await userEvent.click(selectedFilterOption);    
    });

    const filterDropdown = screen.getByTestId('size-option-50');
      
    fireEvent.click(filterDropdown);  
    selectedFilterOption = screen.getByTestId('selected-size-option');

    expect(globalContext.dispatch).toBeCalledWith(actionPagingUpdated);
  });

  it('should select page size 100', async () => {
    render(
    <TodosContext.Provider value={globalContext.state}>
      <TodosDispatchContext.Provider value={globalContext.dispatch} >
        <PageSize inputSelectRef={inputSelectRef} pageCount={2} pageSize={5} totalCount={6} activePage={1} />
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>);
    const actionPagingUpdated = {
      type: TodoActions.pagingUpdated,
      payload: {
        activePage: 1,
        itemsPerPage: 100
      }
    } as IAction;

    let selectedFilterOption = screen.getByTestId('selected-size-option');

    await act(async () => {      
      return await userEvent.click(selectedFilterOption);    
    });

    const filterDropdown = screen.getByTestId('size-option-100');
      
    fireEvent.click(filterDropdown);  
    selectedFilterOption = screen.getByTestId('selected-size-option');

    expect(globalContext.dispatch).toBeCalledWith(actionPagingUpdated);
  });
});