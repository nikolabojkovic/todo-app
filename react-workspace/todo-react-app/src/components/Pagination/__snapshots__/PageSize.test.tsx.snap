// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSize component should match snapshot page size 10 1`] = `
Array [
  <div
    className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center col-sm-3"
  >
    Page 
    1
     of 
    1
  </div>,
  <div
    className="p-3 pt-2 pb-2 col-sm-4"
  >
    <div
      className="justify-content-center hstack gap-2"
    >
      <div
        className="ms-sm-auto"
      >
        Page size:
         
      </div>
      <div
        className=""
      >
        <div
          className="dropdown"
        >
          <button
            aria-expanded={false}
            className="action-button dropdown-toggle btn btn-outline-secondary"
            data-testid="selected-size-option"
            disabled={false}
            id="dropdown-basic"
            onClick={[Function]}
            type="button"
          >
             
            10
             
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`PageSize should display default page size 1`] = `
Array [
  <div
    className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center col-sm-3"
  >
    Page 
    1
     of 
    1
  </div>,
  <div
    className="p-3 pt-2 pb-2 col-sm-4"
  >
    <div
      className="justify-content-center hstack gap-2"
    >
      <div
        className="ms-sm-auto"
      >
        Page size:
         
      </div>
      <div
        className=""
      >
        <div
          className="dropdown"
        >
          <button
            aria-expanded={false}
            className="action-button dropdown-toggle btn btn-outline-secondary"
            data-testid="selected-size-option"
            disabled={false}
            id="dropdown-basic"
            onClick={[Function]}
            type="button"
          >
             
            5
             
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
